omcb.py

```diff
@@ -346,12 +392,93 @@ def generate_img_name(description):
+
+    dates = date_range(start, end)
+    image_count = 0
+    start_string = start.strftime("%m/%d %H:%M:%S")
+    end_string = end.strftime("%m/%d %H:%M:%S")
+    start_seconds = start.timestamp()
+    end_seconds = end.timestamp()
+    total_seconds = end_seconds - start_seconds
+
+    def percent_progress(time):
+        since_start = time.timestamp() - start_seconds
+        percent = since_start / total_seconds
+        return f"{percent: 7.02%}"
+
+    state = None
+    prev_era = None
+    tmpdirname = "./"
+
+    print(f"writing data to {tmpdirname}")
+    diff = blank_int_snapshot()
+    try:
+        snapshot = get_snapshot_name_for_date(dates[0], data_path=data_path)
+        old = state_of_snapshot(snapshot) # initialize first diff to no value
+    except:
+        raise ValueError("Should have at least two states for heatmap mode")
+    for date in dates:
+        era = get_era_for_date(date)
+        if era != prev_era:
+            # We have new snapshots for each era, most relevant for sunsetting because
+            # I wiped the whole grid - need to be careful to load the new snapshot
+            # and we also want to reset our "last snapshot time" so that our timelapse
+            # doesn't have a bunch of dead time during downtime between eras
+            print(f"\nbegin {era} {date}")
+            timelapse_strategy.reset_date_for_new_era()
+            snapshot = get_snapshot_name_for_date(date, data_path=data_path)
+            state = state_of_snapshot(snapshot)
+            diff = add_to_snapshot(state, old, diff)
+            old = state
+        if state is None:
+            snapshot = get_snapshot_name_for_date(date, data_path=data_path)
+            state = state_of_snapshot(snapshot)
+            diff = add_to_snapshot(state, old, diff)
+            old = state
+        log_name = get_log_name_for_date(date, data_path=data_path)
+        with open(log_name, "r") as f:
+                for line in f:
+                    state, date, status = apply_line_to_state_heatmap(state, line, after=start, before=end)
+                    if status == "before-first":
+                        continue
+                    elif status == "past-last":
+                        break
+                    elif status == "continue":
+                        dont_increment_count = False
+                        while timelapse_strategy.should_snapshot_line(date, dont_increment_count):
+                            current_string = date.strftime("%m/%d %H:%M:%S")
+                            progress = percent_progress(date)
+                            description = f"{progress} ({start_string} | {current_string} | {end_string})"
+                            print("\33[2K\r" + description, end="")
+                            state = state_of_snapshot(snapshot)
+                            diff = add_to_snapshot(state, old, diff)
```

Comment:
So the heatmap code is substantially slower than the timelapse code, and I think it's because of the logic here.

With the timelapse code, we apply each line to our state (regardless of the timelapse strategy provided) and then only choose to create an image every `i` seconds or every `n` lines. This saves time because image generation is somewhat expensive (less so with your logic!) and we only call our image gen code when a snapshot is needed.

With the heatmap code, your strategy for updating the heatmap has 1 million instructions (a comparison for each value in the old vs new state). So assuming that we haven't processed 1 million lines between two snapshots (which I think is a reasonable assumption on net), I think we're actually *slower* here than if we just update our heatmap values on every line.

Additionally, I think by only updating our heatmap every `i` seconds (or `n` lines) we're losing some data, since if (say) a box was toggled in both directions several times over those `i` seconds, we won't capture that?

Maybe I'm missing something, but I think a simpler / faster solution here is to fully remove `timelapse_strategy` from the heatmap code and just maintain a mapping `cell_number -> (current_value, number_of_times_changed)` and to update that mapping for each line.

Does this make sense? I can play around with a little patch for this.

Reply:
Maybe it's a little more complicated when you transition to a new day or era and the state of the site fully changes, but it's not actually clear to me that the heatmap should reflect, say, the fact that I wiped the grid near the end.